packages/smart-router/evm/v4-router/queries/getV3Pools.ts

```diff
@@ -0,0 +1,119 @@
+import { ChainId } from '@pancakeswap/chains'
+import { Currency } from '@pancakeswap/sdk'
+import { FeeAmount, TICK_SPACINGS, Tick } from '@pancakeswap/v3-sdk'
+import { multicallByGasLimit } from '@pancakeswap/multicall'
+import { encodeFunctionData, decodeFunctionResult, Address } from 'viem'
+
+import { OnChainProvider, V3Pool } from '../../v3-router/types'
+import { getV3PoolsWithoutTicksOnChain } from '../../v3-router/providers'
+import { tickLensAbi } from '../../abis/ITickLens'
+import { V3_TICK_LENS_ADDRESSES } from '../../constants'
+import { getPairCombinations } from '../../v3-router/functions'
+
+type WithClientProvider = {
+  clientProvider?: OnChainProvider
+}
+
+export type GetV3CandidatePoolsParams = {
+  currencyA?: Currency
+  currencyB?: Currency
+} & WithClientProvider
+
+export async function getV3CandidatePools({ currencyA, currencyB, clientProvider }: GetV3CandidatePoolsParams) {
+  const pairs = getPairCombinations(currencyA, currencyB)
+  return getV3Pools({ pairs, clientProvider })
+}
+
+export type GetV3PoolsParams = {
+  pairs?: [Currency, Currency][]
+} & WithClientProvider
+
+export async function getV3Pools({ pairs, clientProvider }: GetV3PoolsParams) {
+  const pools = await getV3PoolsWithoutTicksOnChain(pairs || [], clientProvider)
+  return fillPoolsWithTicks({ pools, clientProvider })
+}
+
+function getBitmapIndex(tick: number, tickSpacing: number) {
+  return Math.floor(tick / tickSpacing / 256)
+}
+
+type GetBitmapIndexListParams = {
+  currentTick: number
+  fee: FeeAmount
+}
+
+function createBitmapIndexListBuilder(tickRange: number) {
+  return function buildBitmapIndexList<T>({ currentTick, fee, ...rest }: GetBitmapIndexListParams & T) {
+    const tickSpacing = TICK_SPACINGS[fee]
+    const minIndex = getBitmapIndex(currentTick - tickRange, tickSpacing)
+    const maxIndex = getBitmapIndex(currentTick + tickRange, tickSpacing)
+    return Array.from(Array(maxIndex - minIndex + 1), (_, i) => ({
+      bitmapIndex: minIndex + i,
+      ...rest,
+    }))
+  }
+}
+
+// only allow 10% slippage
+const buildBitmapIndexList = createBitmapIndexListBuilder(1000)
```

Comment:
It's making assumptions we can only quote  max 10% slippage route right?

Reply:
yes

Reply:
but it only applies to v3. If there's viable route through v2/stableswap pools whose slippage is greater than 10% we can still route through those pools